libminifi/src/core/Processor.cpp

```diff
@@ -277,6 +286,28 @@ void Processor::onTrigger(const std::shared_ptr<ProcessContext> &context, const
   }
 }
 
+void Processor::callOnTrigger(const std::shared_ptr<ProcessContext>& context, const std::shared_ptr<ProcessSession>& session) {
+  uint64_t num_incoming_flow_files_queued_before = numIncomingFlowFilesQueued();
+  onTrigger(context, session);  // virtual method overridden by subclasses
+  uint64_t num_incoming_flow_files_queued_after = numIncomingFlowFilesQueued();
+
+  if (num_incoming_flow_files_queued_before != 0 && num_incoming_flow_files_queued_before == num_incoming_flow_files_queued_after) {
```

Comment:
As we don't hold lock on the incoming connection during the whole ontrigger execution, flowfiles can be added there in the meanwhile. This condition can be satisfied during normal operation.

Reply:
Yes, good point -- I'll need to rethink this.

Reply:
I have rewritten the code to use a different approach: now `isWorkAvailable()` returns false when all the incoming flow files are penalized.  This will cause the processor to yield for a `nifi.bored.yield.duration` amount of time, so it will no longer spin in a tight loop.